iceoryx_posh/include/iceoryx_posh/internal/roudi/service_registry.hpp

```diff
@@ -30,24 +33,44 @@ namespace roudi
 class ServiceRegistry
 {
   public:
-    static constexpr uint32_t MAX_INSTANCES_PER_SERVICE = 100u;
-    using CaproIdString_t = capro::IdString_t;
-    using InstanceSet_t = cxx::vector<CaproIdString_t, MAX_INSTANCES_PER_SERVICE>;
-    struct instance_t
+    enum class ServiceRegistryError
     {
-        InstanceSet_t instanceSet;
+        INVALID_STATE,
+        SERVICE_DESCRIPTION_ALREADY_ADDED,
+        SERVICE_REGISTRY_FULL,
     };
-    using serviceMap_t = std::map<CaproIdString_t, instance_t>;
+    /// @todo #415 should be connected with iox::MAX_NUMBER_OF_SERVICES
+    static constexpr uint32_t MAX_SERVICE_DESCRIPTIONS = 100U;
+    using CaproIdString_t = capro::IdString_t;
+    using ServiceDescriptionVector_t = cxx::vector<capro::ServiceDescription, MAX_SERVICE_DESCRIPTIONS>;
+
+    /// @brief Adds given service description to registry
+    /// @param[in] serviceDescription, service to be added
+    /// @return ServiceRegistryError, error wrapped in cxx::expected
+    cxx::expected<ServiceRegistryError> add(const capro::ServiceDescription& serviceDescription);
+
+    /// @brief Removes given service description from registry
+    /// @param[in] serviceDescription, service to be removed
+    /// @return true, if service description was removed, false otherwise
+    bool remove(const capro::ServiceDescription& serviceDescription);
+
+    /// @brief Removes given service description from registry
+    /// @param[in] searchResult, reference to the vector which will be filled with the results
+    /// @param[in] service, string or wildcard to search for
+    /// @param[in] service, string or wildcard to search for
+    void find(ServiceDescriptionVector_t& searchResult,
```

Comment:
It may be easier more natural to use like this, but consider the case where we can store 1000 services or more.

There is a reason to have an API like this with our memory constraints, and this is that the return type would always need to support maximum number of elements in the returned container. We have no truly dynamic containers and cannot easily (or at all) define those. It is possible that the user may define the return type as a template parameter, but this is also bad to use for return values (cannot be inferred, needs to be specified explicitly).

Reply:
As @MatthiasKillat in general I agree with you @elfenpiff a `auto result = registry.find("bla", "blubb", "fuu")` is much nicer. A typical ADAS or robotics system will for sure have something around 1000 services. We would not let all the services be findable, but let's assume a size of 200. That's a already a huge number and it will have some runtime effect as we would return lots of `cxx::string`'s via copy here. I'd propose to do a measurement once all building blocks are in place for the new discovery API. This can easily be changed as it is no user-facing API.

Reply:
I like the approach with the lambda. That's quite modern and gives the user maximum freedom